File: typescript/src/bech32m.ts
// Bech32m encoder/decoder (BIP-350), checksum constant 0x2bc830a3.

const CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const CONST = 0x2bc830a3;
const GEN = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];

export function encode(hrp: string, data: Uint8Array): string {
  const values = convertBits(Array.from(data), 8, 5, true);
  if (values === null) {
    throw new Error("cannot regroup payload bits");
  }
  const combined = values.concat(createChecksum(hrp, values));
  return hrp + "1" + combined.map((v) => CHARSET[v]).join("");
}

/** Verify structure + checksum only; returns [hrp, raw 5-bit groups] or null. */
export function decodeRaw(s: string): [string, number[]] | null {
  if (s !== s.toLowerCase() && s !== s.toUpperCase()) {
    return null;
  }
  const lower = s.toLowerCase();
  const pos = lower.lastIndexOf("1");
  if (pos < 1 || lower.length - pos - 1 < 6) {
    return null;
  }
  const hrp = lower.slice(0, pos);
  const dataPart = lower.slice(pos + 1);
  const values: number[] = [];
  for (const ch of dataPart) {
    const idx = CHARSET.indexOf(ch);
    if (idx < 0) {
      return null;
    }
    values.push(idx);
  }
  if (polymod(hrpExpand(hrp).concat(values)) !== CONST) {
    return null;
  }
  return [hrp, values.slice(0, values.length - 6)];
}

/** Decode into [hrp, payload bytes] — the byte-aligned form addresses use. */
export function decode(s: string): [string, Uint8Array] | null {
  const raw = decodeRaw(s);
  if (!raw) {
    return null;
  }
  const bytes = convertBits(raw[1], 5, 8, false);
  return bytes === null ? null : [raw[0], Uint8Array.from(bytes)];
}

// --- internals -----------------------------------------------------------

function polymod(values: number[]): number {
  let chk = 1;
  for (const v of values) {
    const b = chk >>> 25;
    chk = ((chk & 0x1ffffff) << 5) ^ v;
    for (let i = 0; i < 5; i++) {
      if ((b >> i) & 1) {
        chk ^= GEN[i];
      }
    }
  }
  return chk >>> 0;
}

function hrpExpand(hrp: string): number[] {
  const out: number[] = [];
  for (const c of hrp) out.push(c.charCodeAt(0) >> 5);
  out.push(0);
  for (const c of hrp) out.push(c.charCodeAt(0) & 31);
  return out;
}

function createChecksum(hrp: string, data: number[]): number[] {
  const values = hrpExpand(hrp).concat(data, [0, 0, 0, 0, 0, 0]);
  const pm = polymod(values) ^ CONST;
  const out: number[] = [];
  for (let i = 0; i < 6; i++) {
    out.push((pm >> (5 * (5 - i))) & 31);
  }
  return out;
}

/** Regroup a bit stream; null on failure. */
function convertBits(data: number[], from: number, to: number, pad: boolean): number[] | null {
  let acc = 0;
  let bits = 0;
  const maxv = (1 << to) - 1;
  const out: number[] = [];
  for (const value of data) {
    if (value < 0 || value >> from !== 0) {
      return null;
    }
    acc = (acc << from) | value;
    bits += from;
    while (bits >= to) {
      bits -= to;
      out.push((acc >> bits) & maxv);
    }
  }
  if (pad) {
    if (bits > 0) {
      out.push((acc << (to - bits)) & maxv);
    }
  } else if (bits >= from || ((acc << (to - bits)) & maxv) !== 0) {
    return null;
  }
  return out;
}

File: rust/src/bech32m.rs
//! Bech32m encoder/decoder (BIP-350), checksum constant 0x2bc830a3.

const CHARSET: &[u8; 32] = b"qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const CONST: u32 = 0x2bc8_30a3;
const GEN: [u32; 5] = [
    0x3b6a_57b2,
    0x2650_8e6d,
    0x1ea1_19fa,
    0x3d42_33dd,
    0x2a14_62b3,
];

pub fn encode(hrp: &str, data: &[u8]) -> String {
    let mut values = convert_bits(
        &data.iter().map(|&b| b as u32).collect::<Vec<_>>(),
        8,
        5,
        true,
    )
    .expect("regroup");
    let checksum = create_checksum(hrp, &values);
    values.extend_from_slice(&checksum);
    let mut s = String::with_capacity(hrp.len() + 1 + values.len());
    s.push_str(hrp);
    s.push('1');
    for v in &values {
        s.push(CHARSET[*v as usize] as char);
    }
    s
}

/// Verify structure + checksum only; returns (hrp, raw 5-bit groups).
pub fn decode_raw(s: &str) -> Option<(String, Vec<u32>)> {
    if s != s.to_lowercase() && s != s.to_uppercase() {
        return None;
    }
    let lower = s.to_lowercase();
    let pos = lower.rfind('1')?;
    if pos < 1 || lower.len() - pos - 1 < 6 {
        return None;
    }
    let hrp = &lower[..pos];
    let data_part = &lower[pos + 1..];
    let mut values = Vec::with_capacity(data_part.len());
    for ch in data_part.bytes() {
        let idx = CHARSET.iter().position(|&c| c == ch)?;
        values.push(idx as u32);
    }
    let mut check = hrp_expand(hrp);
    check.extend_from_slice(&values);
    if polymod(&check) != CONST {
        return None;
    }
    values.truncate(values.len() - 6);
    Some((hrp.to_string(), values))
}

/// Decode into (hrp, payload bytes) — the byte-aligned form addresses use.
pub fn decode(s: &str) -> Option<(String, Vec<u8>)> {
    let (hrp, values) = decode_raw(s)?;
    let bytes = convert_bits(&values, 5, 8, false)?;
    Some((hrp, bytes.into_iter().map(|v| v as u8).collect()))
}

// --- internals -----------------------------------------------------------

fn polymod(values: &[u32]) -> u32 {
    let mut chk: u32 = 1;
    for &v in values {
        let b = chk >> 25;
        chk = ((chk & 0x1ff_ffff) << 5) ^ v;
        for (i, &g) in GEN.iter().enumerate() {
            if (b >> i) & 1 == 1 {
                chk ^= g;
            }
        }
    }
    chk
}

fn hrp_expand(hrp: &str) -> Vec<u32> {
    let mut out = Vec::with_capacity(hrp.len() * 2 + 1);
    for b in hrp.bytes() {
        out.push((b >> 5) as u32);
    }
    out.push(0);
    for b in hrp.bytes() {
        out.push((b & 31) as u32);
    }
    out
}

fn create_checksum(hrp: &str, data: &[u32]) -> Vec<u32> {
    let mut values = hrp_expand(hrp);
    values.extend_from_slice(data);
    values.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
    let pm = polymod(&values) ^ CONST;
    (0..6).map(|i| (pm >> (5 * (5 - i))) & 31).collect()
}

/// Regroup a bit stream; `None` on failure.
fn convert_bits(data: &[u32], from: u32, to: u32, pad: bool) -> Option<Vec<u32>> {
    let mut acc: u32 = 0;
    let mut bits: u32 = 0;
    let maxv: u32 = (1 << to) - 1;
    let mut out = Vec::new();
    for &value in data {
        if (value >> from) != 0 {
            return None;
        }
        acc = (acc << from) | value;
        bits += from;
        while bits >= to {
            bits -= to;
            out.push((acc >> bits) & maxv);
        }
    }
    if pad {
        if bits > 0 {
            out.push((acc << (to - bits)) & maxv);
        }
    } else if bits >= from || ((acc << (to - bits)) & maxv) != 0 {
        return None;
    }
    Some(out)
}
